### DISCUSSION TITLE
How to pack the fyne app  and *so together?

### DISCUSSION THREAD

**Discussion by fayfive** [fyne-io/fyne#3367] [Q&A] [answered]
I have use fyne in my app. But when I `go build -ldflags '-w -s -linkmode "external" -extldflags "-static"'`
It can not be found `-lGL`, but I have `apt install` and find the `/usr/lib/x86_64-linux-gnu/libGL.so`.
`go build` is fine. But I want to pack the *.so together

**Comment by andydotxyz** [comment, marked as answer]
We don't support building a Fyne app in this way. Bundling the GL driver into the app will likely break on certain runtimes as it varies based on your graphics card.
What is your aim in taking this approach?

**Reply by fayfive** [reply]
Because I want to run without any depends on the other PC.

**Reply by andydotxyz** [reply]
I don't quite understand this. If there is a graphics output there should be a graphics driver. If there is no graphics driver then bundling it into the application won't help as it won't be able to connect to a graphics output device?

**Reply by fayfive** [reply]
Because some library of my system is too high. Similarly libc.so.6 or libc++.so.6. It can not run on the lower libc system.
Any suggestions?
Because I can do it when I run my back end app what is written by golang.

**Reply by andydotxyz** [reply]
The easiest way to support lower libc is to build on an older system, another way is to use fyne-cross which has a sandardised build for different platforms which may help.
We are looking at alternatives to libc on linux which would enable a "more static" build where it is literally just the GL driver required, but that has various challenges and draw backs as well. 
It sounds like an intriguing system you are building for - what sort of a computer cannot install a recent libc implementation?

**Reply by fayfive** [reply]
Some user‘s PC is very old, maybe the Linux kernel is version 3. That is one point. I just want to all lib in the app similarly windows.

**Reply by Bluebugs** [reply]
The GL is, as @andydotxyz pointed out, per GPU and kernel version. It is not possible on Linux to ship a single GL library that will cover all GPU and if you add to this old kernel, it won't work. Now, if your application only need to support one specific GPU and a specific kernel, it might be possible. What configuration do you need to support?

**Reply by fayfive** [reply]
I know the GL is about the GPU. It must be for the machine. I don't want to the users to install something when they run my app.

**Reply by andydotxyz** [reply]
We agree. The design is that no additional dependencies are required. If their OS is up to date it should work out of the box.

**Reply by fayfive** [reply]
I know the newest version will work well. But on Linux, we are difficult to make sure all the users have update to the newest version. As I know, libc of Ubuntu18.04 and 22.04 are different, many persons are still working on the ubuntu 18.04.
So I want to know if I can package all the *.so what the app needs except GL.so when I compile? Or use some tools to do it?

**Reply by Jacalz** [reply]
As far as I know, it is not possible to package `glibc` statically into a binary. You would have to use `musl` but that is going to bring up a whole host of other issues (not even sure if glfw works with it). I'd say that the likely answer to that question is no.

**Reply by andydotxyz** [reply]
Your best bet is to compile on the oldest Linux you want to support, as it will be forwards compatible.

**Reply by fayfive** [reply]
@Jacalz I can build it with `go build -ldflags '-w -s -linkmode "external" -extldflags "-static"'`. The back end app with  "C" can use on all ubuntu or centos.

**Reply by fayfive** [reply]
@andydotxyz Can I use Appimage to package?

**Reply by Jacalz** [reply]
There should not be anything preventing you from using AppImage. You can also use Flatpak if you prefer.

**Reply by andydotxyz** [reply]
> @Jacalz I can build it with `go build -ldflags '-w -s -linkmode "external" -extldflags "-static"'`. The back end app with  "C" can use on all ubuntu or centos.

The backend does not have to connect to graphics. We are not saying it isn't possible with Go in general, but when you add graphics everything gets more complicated - so we are trying to describe what is possible for GUI app compiling.
